ScilifelabDataCentre/dds_web
# Change argon2 password settings

## Before submitting this PR

1. **Description:** 
    - At the moment, the argon2 settings for the passwords differ from those used when getting the private key. 
    - This PR defines the argon2 settings in config.py: It's not possible to change most settings for the key derivation during retrieval of the private key, but it's possible to set the password settings. If not, they are defaulted to the same as the key derivation.
    - This means that we can try out the different settings for the passwords in an easy way.
    - This should not break anything, since the passwords for all users should be automatically rehashed if the settings are changed in any way. 
        1. User is created during registration: Password is hashed and the full settings and password hash is stored in `user._password_hash` 
        2. User authenticates: No hash is needed because the PasswordHasher uses the settings defined in the password hash; it's not hard coded in the code.
        3. The settings change in some way. 
        4. User authenticates: Authentication still works because nothing is hard coded and the settings are still taken from the hash itself. Rehash is needed though, because the PasswordHasher notices a setting change. `verify_password` then automatically rehashes and saves the new hash. 
2. **Jira task / GitHub issue:** Part of DDS-1439
3. **How to test:**
    1. Start up the api with docker 
    2. Authenticate with `dds auth login`
    3. Change one of the settings in `models.py` in `User.verify_password()`, e.g. `hash_len`
    4. Authenticate again as in step 2 - rehashing should work 
    5. Authenticate again as in step 2 - no rehashing should be done 
    6. Try to download data - should work 
7. **Type of change:** [_Check the relevant boxes in the section below_](#what-type-of-changes-does-the-pr-contain)
8. **Add docstrings and comments to code**, _even if_ you personally think it's obvious.

## What _type of change(s)_ does the PR contain?




## Checklist

- [Sprintlog](../SPRINTLOG.md)
- Rebase / Update / Merge _from_ base branch (the branch from which the current is forked)
- Blocking PRs
- PR to `master` branch

## Actions / Scans



- **Black**: Python code formatter. Does not execute. Only tests.
  Run `black .` locally to execute formatting.
- **Prettier**: General code formatter. Our use case: MD and yaml mainly.
  Run `npx prettier --write .` locally to execute formatting.
- **Yamllint**: Linting of yaml files.
- **Tests**: Pytest to verify that functionality works as expected.
- **CodeQL**: Scan for security vulnerabilities, bugs, errors
- **Trivy**: Security scanner
- **Snyk**: Security scanner


<a href="https://gitpod.io/#https://github.com/ScilifelabDataCentre/dds_web/pull/1406"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
